### PR TITLE
feat: user apps

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

https://github.com/discord/discord-api-docs/commit/1eaa1af5677712957479c9959132003b79396e24 :woo:
(+ https://github.com/discord/discord-api-docs/commit/462cddc4253f1d46bc05fb7d6809b94a9d2f118d)

Adds support for all things related to user-installed applications.
Some parts of the API aren't... great. Particularly some of the terminology/naming used is just confusing, and some parts are just completely nontransparent wrt. defaults and fields taking priority over others. It is what it is. /shrug

#### Issues
A probably non-exhaustive list of things that are still missing, could be improved, or are still uncertain.

API issues/unclear behavior:

1. ~~`contexts` does not default to `[0, 1, 2]` on the api side, despite the documentation's claims.~~
    - resolved: this shouldn't matter
2. It's not clear how `Application.install_params` and `.integration_types_config` work together. Right now it seems that the latter takes precedence over the former, if set. The dev portal sends both, looks like the API isn't doing any data shuffling here.
3. ~~Same thing goes for `ApplicationCommand.dm_permission` and `.contexts`. I'm not sure if we even want to deprecate `dm_permission` right now.~~
    - resolved: deprecated `dm_permission`; if both are set, we raise an exception. fwiw, if both are sent, the API prefers `contexts` over `dm_permission`
4. ~~We might want to send a fixed default value for `contexts` instead of the current `null`, in case the API ever starts choosing for us.~~
    - resolved: likely not needed, and can be fixed trivially if it ever happens. worse case is that we sync unnecessarily
5. ~~The `authorizing_integration_owners` field is weird, and its structure varies heavily based on configuration/context of commands and interactions. It already has a [lengthy docstring](https://github.com/DisnakeDev/disnake/blob/545268b8a499437e01781d6ee781ca01f30fe767/disnake/interactions/application_command.py#L93-L110), but it might be easier to instead split it into two separate properties and add some of our own fallbacks.~~
    - resolved: the API somewhat improved here, the fields now seems to be provided in basically all situations; the data is part of a new `AuthorizingIntegrationOwners` class now
6. ~~`message.interaction_metadata.name` isn't documented - quote, "well it will still be there, but i dont want to officially document it". It's an optional field, so implementing it nonetheless is probably fine.~~
    - resolved: decided to remove this attribute

Bugs/Missing features:

7. ~~This needs #1012, #1233. Additionally, #1236 for the `guild` attribute in guilds without the bot user. And then likely a ton of mocking for `guild.me` and several permissions-related methods. Still very uncertain about this part.~~
    - (partially) resolved: all prerequisite PRs have been merged, except the guild one, which I don't plan on including for now, since it has lots of edge cases and potentially far-reaching consequences
8. ~~The `GuildCommandInteraction` annotation (which ends up setting `dm_permission=False`) silently conflicts with the `contexts` parameter. See 3.~~
    - resolved: the annotation now sets `contexts = InteractionContextTypes(guild=True)` instead of `dm_permission = False`

QOL:

9. Bot-wide defaults for the two new command fields could be useful, but aren't trivial to implement. This will be a separate PR.
10. ~~There are no utility decorators yet (like `@commands.default_member_permissions()`).~~
    - resolved: added `@commands.integration_types` and `@commands.contexts` decorators
11. ~~Allowing a command e.g. in all contexts is very verbose due to `contexts=[InteractionContextType.guild, ..., ...]`. An `.all()` property on the two new enums could be useful, but also feels a bit weird. Might be resolved by 10. eventually.~~
    - resolved: the enums were turned into flags, which also have `.all()`
12. ~~As mentioned, the official naming is weird. "interaction context" vs "installation context"/"integration type" is easy to mix up, and "integration type" in itself just seems... wrong. Considering renaming all `integration_types` to `installation_types` (the *one* combination of these words that the api docs *don't* use!).~~
    - resolved: the official naming remains weird. all "integration types" have been renamed to "install types", which feels like the best description of that particular field.
13. Another followup PR is likely needed to cover smaller issues related to `Interaction.guild`, e.g. in ext.commands checks.
14. More naming stuff can be found at various `TODO`s in the code.

Documentation:

15. ~~The [docs page](https://docs.disnake.dev/en/latest/ext/commands/slash_commands.html#permissions) still needs updates.~~
    - resolved: updated the documentation to cover `integration_types` and `contexts`, and removed the `dm_permission` part

## Checklist

- [x] If code changes were made, then they have been tested
    - [ ] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `pdm lint`
    - [x] I have type-checked the code by running `pdm pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by DJStompZone** [comment]
> 12. As mentioned, the official naming is weird. "interaction context" vs "installation context"/"integration type" is easy to mix up, and "integration type" in itself just seems... wrong. Considering renaming all `integration_types` to `installation_types` (the _one_ combination of these words that the api docs _don't_ use!).

I agree that the naming is weird, and it still may change. But personally I feel like naming things similarly but differently compared to the discord docs could just wind up adding unnecessary confusion to an already murky situation. Especially for people trying to search by specific keywords.

File: disnake/message.py [hunk 1]
```diff
@@ -714,6 +726,77 @@ def author(self) -> User:
...
+        "original_response_message_id",
+        "name",
+        "interacted_message_id",
+        "triggering_interaction_metadata",
+    ]
+
+    def __init__(self, *, state: ConnectionState, data: InteractionMetadataPayload) -> None:
+        self._state: ConnectionState = state
+
+        self.id: int = int(data.get("id"))
```
[the remark this answers is not in the document]
Comment (shiftinv) [reply disnake/message.py]: > any reason why you're not doing

nope, no reason.
I wanted to blame this on writing code late at night, but apparently this happened at 3pm, so I have no excuses - thanks for the heads up :>

https://github.com/DisnakeDev/disnake/pull/1173/commits/10a8bbbafb005d4979b08d7dfa2a508bc4d6f057

**Comment by onerandomusername** [comment]
DM Permission requires a bit more work (shouldn't impact user-facing side, should not actually touch the API) but this was the quickest change to make it no longer touch the API

**Comment by onerandomusername** [comment]
@shiftinv could you separate the stuff for app info (disnake/appinfo.py, disnake.utils.oauth_url, etc) into a separate pr? Those don't need to be a part of this PR.

**Comment by shiftinv** [comment]
Going to go ahead and merge this now; I'm quite happy with the current state, and it's been long overdue.

The only two remaining points are bot-wide defaults for `install_types` and `contexts`, plus some potential adjustments related to `Interaction.guild` in ext.commands checks. These aren't crucial, and will be added in a future (as in, likely within a couple days) PR.

**Pull request merged**